### PR TITLE
Dynamic Indexed Expression Generation

### PR DESCRIPTION
**Pull request by kennethsible**
- Implemented declare_indexedexp() function for dynamically generating an indexed expression with NRPy+ notation using Python metaprogramming.
- Changed behavior of the following functions to wrapper functions of declare_indexedexp(): zerorank1(), zerorank2(), zerorank3(), zerorank4(), declarerank1(), declarerank2(), declarerank3(), declarerank4()

### CODE REVIEW

File: indexedexp.py [hunk 1]
```diff
@@ -10,25 +10,88 @@
...
+                        elif symmetry == 'sym12':
+                            if k < j: indexedexp[i][j][k] = indexedexp[i][k][j]
+                        elif symmetry == 'nosym': pass
+                        else: raise Exception('unsupported symmetry option \'' + symmetry + '\'')
+        elif rank == 4:
+            for i in range(dimension):
+                for j in range(dimension):
+                    for k in range(dimension):
+                        for l in range(dimension):
+                            if symmetry in ('sym01', 'sym01_sym23'):
```
Comment (zachetienne) [inline indexedexp.py:74]: We are missing 'sym23' as an option here

Comment (zachetienne) [reply indexedexp.py:74]: Technically 'sym02', 'sym03', and 'sym13' should also be included

File: indexedexp.py [hunk 2]
```diff
@@ -10,25 +10,88 @@
...
+                for j in range(dimension):
+                    if symmetry == 'sym01':
+                        if j < i: indexedexp[i][j] = indexedexp[j][i]
+                    elif symmetry == 'nosym': pass
+                    else: raise Exception('unsupported symmetry option \'' + symmetry + '\'')
+        elif rank == 3:
+            for i in range(dimension):
+                for j in range(dimension):
+                    for k in range(dimension):
+                        if symmetry == 'sym01':
```
Comment (zachetienne) [inline indexedexp.py:63]: Technically 'sym02' should also be included

**Comment by kennethsible** [comment]
I added a symmetry option for every index permutation and condensed the symmetry code using the itertools package.

**Pull request merged**